File: examples/client/stalemate_expire.py
"""Flow C — stalemate → expiry refund.

Client disputes but quorum is never reached. Settlement is impossible
because the policy stays PENDING. Once the job passes its ``expiredAt``
deadline, the client (or anyone) calls ``claimRefund`` — the universal
escape hatch on the kernel.

NOTE: ``expiredAt`` must be > jobExpiry; the demo uses a very small
``jobExpirySeconds`` so it completes during a run. For live networks set
``expired_at`` to something reasonable (hours / days).
"""

from __future__ import annotations

import time

from _helpers import banner, load_settings, make_client, minutes_from_now

from bnbagent.apex import DeliverableManifest, JobStatus, SCHEMA_VERSION


def main() -> None:
    s = load_settings()
    client = make_client(s.client_pk, s.network)

    banner("STALEMATE — dispute without quorum, refund at expiry")

    decimals = client.token_decimals()
    budget = 1 * (10 ** decimals)

    # Short expiry so the demo finishes in-session. Still >= the 5-minute
    # on-chain minimum; bump higher for mainnet.
    expired_at = minutes_from_now(6)

    res = client.create_job(
        provider=s.provider_address,
        expired_at=expired_at,
        description="APEX demo: stalemate",
    )
    job_id = res["jobId"]
    print(f"[client] createJob jobId={job_id} expiredAt={expired_at}")
    client.register_job(job_id)
    client.set_budget(job_id, budget)
    client.fund(job_id, budget)

    if not s.provider_pk:
        print(f"\nProvider must submit jobId={job_id} before continuing.\n")
        return
    provider = make_client(s.provider_pk, s.network)
    manifest = DeliverableManifest(
        version=SCHEMA_VERSION,
        job_id=job_id,
        chain_id=provider.network.chain_id,
        contracts={
            "commerce": provider.commerce.address,
            "router": provider.router.address,
            "policy": provider.policy.address,
        },
        response={"content": f"stalemate test result for job {job_id}", "content_type": "text/plain"},
        submitted_at=0,  # back-filled with block.timestamp after submit
    )
    # In production: upload manifest.to_dict() to IPFS/storage first, then pass the URL.
    # deliverable_url = storage.upload(manifest.to_dict(), f"job-{job_id}.json")
    deliverable_url = ""  # no storage in this example
    receipt = provider.submit(job_id, manifest.manifest_hash(), {"deliverable_url": deliverable_url})
    manifest.submitted_at = provider.w3.eth.get_block(receipt["blockNumber"])["timestamp"]
    # In production: re-upload manifest.to_dict() now that submitted_at is set.
    print("[provider] submit OK")

    client.dispute(job_id)
    print("[client] dispute raised (no voter will act)")

    wait = expired_at - int(time.time()) + 3
    if wait > 0:
        print(f"[client] waiting {wait}s for expiry...")
        time.sleep(wait)

    client.claim_refund(job_id)
    job = client.get_job(job_id)
    assert job.status == JobStatus.EXPIRED, f"expected EXPIRED, got {job.status.name}"
    print(f"[client] claimRefund OK -> {job.status.name}")


if __name__ == "__main__":
    main()

File: typescript/examples/client/stalemateExpire.ts
/**
 * Flow C — stalemate -> expiry refund.
 *
 * Client disputes but quorum is never reached. Settlement is impossible
 * because the policy stays PENDING. Once the job passes its `expiredAt`
 * deadline, the client (or anyone) calls `claimRefund` — the universal
 * escape hatch on the kernel.
 *
 * NOTE: `expiredAt` must accommodate the policy's dispute window because the
 * provider has to submit before `expiredAt - disputeWindow`. The script's
 * wall-clock wait scales with that window, so on a network configured with a
 * multi-hour window this demo is slow by design.
 *
 * Port of `python/examples/client/stalemate_expire.py`. Hits live testnet —
 * run manually, not in CI.
 */

import {
  DeliverableManifest,
  JobStatus,
  SCHEMA_VERSION,
} from "../../src/erc8183/index.js";
import {
  banner,
  expiryFor,
  loadSettings,
  makeClient,
  makePrimaryClient,
  sleep,
} from "./_helpers.js";

async function main(): Promise<void> {
  const s = loadSettings();
  const client = await makePrimaryClient(s);

  banner("STALEMATE — dispute without quorum, refund at expiry");

  const decimals = await client.tokenDecimals();
  const budget = 1n * 10n ** BigInt(decimals);

  // Smallest expiry that still admits a valid submit: disputeWindow + 1 min.
  const expiredAt = await expiryFor(client, 1);

  const res = await client.createJob({
    provider: s.providerAddress,
    expiredAt,
    description: "ERC-8183 demo: stalemate",
  });
  const jobId = res.jobId;
  if (jobId === null) {
    throw new Error("createJob did not return a jobId");
  }
  console.log(`[client] createJob jobId=${jobId} expiredAt=${expiredAt}`);
  await client.registerJob(jobId);
  await client.setBudget(jobId, budget);
  await client.fund(jobId, budget);

  if (!s.providerPk) {
    console.log(`\nProvider must submit jobId=${jobId} before continuing.\n`);
    return;
  }
  const provider = await makeClient(s.providerPk, s.network);
  const manifest = new DeliverableManifest({
    version: SCHEMA_VERSION,
    jobId: Number(jobId),
    chainId: provider.network.chainId,
    contracts: {
      commerce: provider.commerce.address,
      router: provider.router.address,
      policy: provider.policy.address,
    },
    response: {
      content: `stalemate test result for job ${jobId}`,
      contentType: "text/plain",
    },
  });
  // In production: upload manifest.toDict() to IPFS/storage first, then pass the URL.
  const deliverableUrl = "https://example.invalid/manifest.json"; // placeholder — these scripts test on-chain flow only
  await provider.submit(jobId, manifest.manifestHash(), {
    deliverable_url: deliverableUrl,
  });
  console.log("[provider] submit OK");

  await client.dispute(jobId);
  console.log("[client] dispute raised (no voter will act)");

  const wait = Number(expiredAt) - Math.floor(Date.now() / 1000) + 3;
  if (wait > 0) {
    console.log(`[client] waiting ${wait}s for expiry...`);
    await sleep(wait * 1000);
  }

  await client.claimRefund(jobId);
  const job = await client.getJob(jobId);
  if (job.status !== JobStatus.EXPIRED) {
    throw new Error(`expected EXPIRED, got ${JobStatus[job.status]}`);
  }
  console.log(`[client] claimRefund OK -> ${JobStatus[job.status]}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
